Min php81 with fixes

mattwithoos opened the pull request.

mattwithoos committed:
drop support for hirak; no compatible version

mattwithoos committed:
patch to 8.1

mattwithoos committed:
update to php81 and fix type safety

mattwithoos committed:
fix tests for php81 upgrade

mattwithoos committed:
change tact and support >=8.0 - <9

mattwithoos committed:
enable actions for first time

david-ethicaljobs commented on src/RequestUtil.php:
```diff
@@ -2,24 +2,26 @@
+    public static function getSelectFields(): array
     {
-        if ($select = request()->input('fields')) {
+        $grossGlobalMethod = request();
```
❤️

david-ethicaljobs reviewed (commented):
nice work, I approve all these changes but Dani made a good point in slack, We prob still need to check this codebase on 7.4 as we won't be able to update all the other repos that depend on this.

00dani committed:
Fix deprecated PHPUnit configuration

00dani committed:
Gitignore PHPUnit cache

00dani committed:
Add PHPStan to help catch errors

00dani committed:
Fix types in ApiResources

00dani committed:
Fix implementation of Arrays::objectoArray

I can only assume this method has never been called, because the
previous implementation absolutely would not work at all. Huh.

00dani committed:
Fix types in Arrays

00dani committed:
Require jetbrains/phpstorm-attributes for PHPStan

00dani committed:
Fix types in Dates

00dani committed:
Fix types in RequestUtil

00dani committed:
Fix types in Money (ugh)

00dani committed:
Fix types in Timestamp

00dani committed:
Clean up flow of Timestamp::toMilliseconds

00dani committed:
Add phpstan-phpunit to help type the tests

00dani committed:
Fix types for Fixtures\Person

00dani committed:
Help narrowing in Timestamp

00dani committed:
Avoid class-string in ApiResources :(

Unfortunately, we can't write this code in a typesafe way and also have
the unit tests work, because the tests are running in this codebase,
where nothing in the \App namespace exists at all. Either we fail to get
a string back from the method at all (if we assert(class_exists())
inside it), or PHPStan correctly points out that the test is trying to
compare a string that does not refer to an existing class name with a
class-string.

So we've just gotta fall back on returning regular strings, sadly.
Sighs.

00dani committed:
Update test suite to typecheck

Also, to be less horrible. A bunch of assertions were written backwards,
with the expected and actual values in each others' places? So I fixed
that. Also fixed the incorrect namespacing of all these test classes,
plus got them all to typecheck, and they all pass with flying colours.

00dani committed:
Tell PHPStan to ignore unreachable skipped tests

00dani committed:
Enable PHPStan in Github Actions

mattwithoos committed:
Update tests.yaml

00dani committed:
Try to teach Github to recognise PHPUnit output

I'm not sure if doing it like this works, but I guess I'll see when it
runs?

00dani committed:
Clean up formatting of Github workflow

00dani committed:
Correctly request bcmath in Github workflow

00dani committed:
Avoid loading Laravel for tests that don't need it

00dani committed:
Use sleight of hand for type-compat with old Carbon

Older versions of Carbon::createFromTimestamp() declare that the
timestamp needs to be an integer, but a numeric-string will actually
work just fine. In newer Carbon, numeric-strings are handled differently
and will support bigger numbers, which we would like to take advantage
of.

We could use a cast to int to placate PHPStan, but that would mean we
lose the ability to work with really big timestamps ourselves. This
approach is compatible with both the oldeer and newer Carbon packages
and will also be approved by PHPStan (because we're basically tricking
it).

00dani committed:
Swap PHPUnit printer so we can lock Orchestra versions

Since this package is a dependency for others, it kinda needs to try to
work with different versions of its dependencies. To that end, in the
Github workflow we install its deps with both --prefer-stable and
--prefer-lowest, which helps test that at least two versions of the deps
work okay.

Unfortunately that meant pulling in a completely different major version
of orchestra/testbench, which was a problem because the older one just
crashed PHPUnit entirely. I guess PHPUnit itself changed?

I couldn't insist on getting version 7 of the Orchestral packages
because they require a newer Symfony YAML package, and the result
printer we were using required older versions of that YAML package. No
idea why it needs a YAML parser to pretty-print PHPUnit results, but
whatever.

This new result printer package doesn't depend on anything and shouldn't
cause further dependency hell. It also looks a little nicer in my
opinion. I like BDD.

00dani committed:
Fix a tiny formatting mistake
